lib/resty/ldap/filter.lua

```diff
@@ -0,0 +1,156 @@
+    local args = pack(...)
+    return {
+        item_type = item_type,
+        attribute_description = args[1].attribute_description,
+        filter_type = args[2].filter_type,
+        attribute_value = args[3].attribute_value,
+    }
+end
+
+-- Simple types
+local rawValue = (P'_' + R('az', 'AZ')) * (P'_' + R '09' + R('az', 'AZ')) ^ 0
+
+-- Grammar
+local filter = P{
+    _'FILTER' / function (f)
+        return f
+    end,
+    FILTER = _'FILL' * P'(' * _'OP' * P')' * _'FILL' / function(f)
+        return f
+    end,
+
+    OP= (_'OP_AND' + _'OP_OR' + _'OP_NOT' + _'ITEM'),
+    OP_AND = (P'&' * _'FILL' * list('FILTER', 1) * _'FILL') / function(...)
+        return cOPBody(_M.OP_TYPE_AND, ...)
+    end,
+    OP_OR = (P'|' * _'FILL' * list('FILTER', 1) * _'FILL') / function(...)
+        return cOPBody(_M.OP_TYPE_OR, ...)
+    end,
+    OP_NOT = 
+        (P'!' * (
+            (_'FILL' * _'FILTER' * _'FILL') +
+            (_'FILL' * _'ITEM' * _'FILL')
+        )) / function(...)
+            return cOPBody(_M.OP_TYPE_NOT, ...)
+        end,
+    ITEM = _'ITEM_SUBSTRING' +  _'ITEM_SIMPLE' + _'ITEM_PRESENT',
+    ITEM_SUBSTRING = (_'ATTRIBUTE_DESCRIPTION' * _'FILTER_TYPE_EQUAL' * _'ATTRIBUTE_VALUE_SUBSTRING') / function(...)
+        return cItemBody(_M.ITEM_TYPE_SUBSTRING, ...)
+    end,
+    ITEM_SIMPLE = _'ATTRIBUTE_DESCRIPTION' * _'FILTER_TYPE' * _'ATTRIBUTE_VALUE' / function(...)
+        return cItemBody(_M.ITEM_TYPE_SIMPLE, ...)
+    end,
+    ITEM_PRESENT = _'ATTRIBUTE_DESCRIPTION' * P'=*' / function (value)
+        return cItemBody(
+            _M.ITEM_TYPE_PRESENT, value,
+            { filter_type = _M.FILTER_TYPE_EQUAL },
+            { attribute_value = "*" }
+        )
+    end,
+    FILTER_TYPE = _'FILTER_TYPE_EQUAL' + _'FILTER_TYPE_APPROX' + _'FILTER_TYPE_GREATER' + _'FILTER_TYPE_LESS',
+    FILTER_TYPE_EQUAL = P'=' / function()
+        return { filter_type = _M.FILTER_TYPE_EQUAL }
+    end,
+    FILTER_TYPE_APPROX = P'~=' / function()
+        return { filter_type = _M.FILTER_TYPE_APPROX }
+    end,
+    FILTER_TYPE_GREATER = P'>=' / function()
+        return { filter_type = _M.FILTER_TYPE_GREATER }
+    end,
+    FILTER_TYPE_LESS = P'<=' / function()
+        return { filter_type = _M.FILTER_TYPE_LESS }
+    end,
+
+    ATTRIBUTE_DESCRIPTION = rawValue / function(value)
+        return { attribute_description = value }
+    end,
+    ATTRIBUTE_VALUE = rawValue / function(value)
+        return { attribute_value = value }
+    end,
+    ATTRIBUTE_VALUE_SUBSTRING =
+        ((_'ATTRIBUTE_VALUE' * _'WILDCARD') +
+        (_'WILDCARD' * _'ATTRIBUTE_VALUE' * _'WILDCARD') +
+        (_'WILDCARD' * _'ATTRIBUTE_VALUE')) / function(...)
+            local args = pack(...)
+            local s = ''
+            for i = 1, args.n, 1 do
+                local value = args[i]
+                if type(value) == "table" then
+                    s = s .. value.attribute_value
+                else
+                    s = s .. value
```

Comment:
Do we need to use table.concat in this function?

Reply:
Optimized

Reply:
@spacewander optimized